### DISCUSSION TITLE
Great tool !!! Thank you.

### DISCUSSION THREAD

**Discussion by cbazza** [mike-lischke/java2typescript#16] [General]
Just wanted to say that I used your tool to convert a large Java project to Typescript (running inside the browser) and it has helped me tremendously.
The biggest issue I had was with circular dependencies which I solved by using dependency injection via an initialization function. My method was pretty mechanical and could be automated. I used a tool (madge) to analyse the dependencies and it showed 62 circular dependencies.

Thank you for this great tool.

Carlos.

**Comment by mike-lischke** [comment]
Great the tool helped you so much! And thanks for your nice feedback.

I also use it in various projects and found that circular dependencies problem almost always the biggest one. I usually use a combination of dependency injection and interfaces. I fact, interfaces have been my universal tool to get the circular dependencies solved.

Finding the circular dependencies is relatively easy, but solving that automatically is something I imagine not so simple, as it requires detailed knowledge of the project.

**Reply by cbazza** [reply]
The thing about my project is that I did not want to change (refactor) the original code or the converted code. So I was doing everything to make the conversion mechanically done without much thinking.
My method is equivalent to using `interfaces` and it goes like this:

1. use `import type` which only loads type and doesn't show up on compiled JS code (equivalent to `interfaces`).
2. define local module variable for injected `Class` that starts with `$`. 
3. modify all local instantiation of that `Class` to use the `$Class` version (eq. `new $Class()`).
4. provide initialization function for all injected classes in the file.
5. wire all up in main `index.js` for that package/folder.

Check out this example where classes `Mult` and `Div` are circular:
**Mult.ts**
```
import type { Sum } from "./Sum";
...
let $Sum = null;
export function injectDepsForMult(Parser, Float_, Unary, Div, Sum, Pwr, Var, Top) {
   ...     
   $Sum = Sum;
}
// Use $Sum instead of Sum in this file
// eq. let sum: Sum = new $Sum();
...
``` 
**Sum.ts**
```
import type { Mult } from "./Mult";

let $Mult = null;
export function injectDepsForSum(Func, Unary, Float_, Div, Mult, Top, Pwr) {
    ...
    $Mult = Mult;
}
// Use $Mult instead of Mult in this file
// eq. let mult: Mult = new $Mult();
...
```

**index.js**
```
import { Mult, injectDepsForMult } from "./Mult";
import { Sum, injectDepsForSum } from "./Sum";
...
injectDepsForMult(Parser, Float_, Unary, Div, Sum, Pwr, Var, Top);
injectDepsForSum(Func, Unary, Float_, Div, Mult, Top, Pwr);

```

**Comment by mike-lischke** [comment]
Interesting approach. Might be helpful for people looking for a solution for their projects. Trying to solve deep circular dependencies can be a daunting task, and made me more than once curse the JS gods for that :-D
